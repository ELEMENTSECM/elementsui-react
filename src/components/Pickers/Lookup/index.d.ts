import * as React from 'react';

export type LookupValue =
	| {
			value?: any;
			label?: string;
	  }
	| {
			value?: any;
			label?: string;
	  }[]
	| Object[];

export interface LookupStyles {
	clearIndicator?: (...args: any[]) => any;
	container?: (...args: any[]) => any;
	control?: (...args: any[]) => any;
	dropdownIndicator?: (...args: any[]) => any;
	group?: (...args: any[]) => any;
	groupHeading?: (...args: any[]) => any;
	indicatorsContainer?: (...args: any[]) => any;
	indicatorsSeparator?: (...args: any[]) => any;
	input?: (...args: any[]) => any;
	loadingIndicator?: (...args: any[]) => any;
	loadingMessage?: (...args: any[]) => any;
	menu?: (...args: any[]) => any;
	menuList?: (...args: any[]) => any;
	multiValue?: (...args: any[]) => any;
	multiValueLabel?: (...args: any[]) => any;
	multiValueRemove?: (...args: any[]) => any;
	noOptionsMessage?: (...args: any[]) => any;
	option?: (...args: any[]) => any;
	placeholder?: (...args: any[]) => any;
	singleValue?: (...args: any[]) => any;
	valueContainer?: (...args: any[]) => any;
}

export type LookupMenuPlacement = 'bottom' | 'top' | 'auto';

export type LookupNoOptionsMessage = ((...args: any[]) => any) | any;

export type LookupLoadingMessage = ((...args: any[]) => any) | any;

export interface LookupProps {
	/**
	 * The id to set on the SelectContainer component
	 */
	id?: string;
	/**
	 * The id of the search input
	 */
	inputId?: string;
	/**
	 * Initial value
	 */
	value?: LookupValue;
	/**
	 * If true, the box will be unselectable, can be changed on the fly
	 */
	disabled?: boolean;
	/**
	 * Callback executed when a value has been selected
	 */
	onChange?: (...args: any[]) => any;
	/**
	 * Renders text for each search result line
	 */
	renderOption?: (...args: any[]) => any;
	/**
	 * Used to retrieve a key from entity record, by default uses Entity.key property (that in turn returns Id)
	 */
	idSelector?: (...args: any[]) => any;
	/**
	 * Function that returns odata query for provied search term
	 */
	queryProvider?: (...args: any[]) => any;
	/**
	 * Used to filter result set before rendering dropdown
	 */
	resultsFilter?: (...args: any[]) => any;
	/**
	 * Scroll pagination. 20 by default
	 */
	pageSize?: number;
	/**
	 * Throttle delay, 300 ms by default
	 */
	delay?: number;
	/**
	 * A custom error message otherwise the general error message will be used
	 */
	errorMessage?: string;
	/**
	 * Reference to select element
	 */
	selectRef?: (...args: any[]) => any;
	/**
	 * Lookup options
	 */
	options?: Object[];
	/**
	 * Placeholder
	 */
	placeholder?: string;
	/**
	 * Root element's class name
	 */
	className?: string;
	/**
	 * Allow the user to select multiple values
	 */
	isMulti?: boolean;
	/**
	 * Custom styles
	 */
	styles?: LookupStyles;
	/**
	 * Theme override
	 */
	theme?: (...args: any[]) => any;
	/**
	 * Value can be cleared
	 */
	isClearable?: boolean;
	/**
	 * Default placement of the menu in relation to the control. 'auto' will flip when there isn't enough space below the control.
	 */
	menuPlacement?: LookupMenuPlacement;
	/**
	 * Text to display when there are no options
	 */
	noOptionsMessage?: LookupNoOptionsMessage;
	/**
	 * Async: Text to display when loading options
	 */
	loadingMessage?: LookupLoadingMessage;
	/**
	 * Include full object value
	 */
	fullObjectValue?: boolean;
	/**
	 * JSX elements to be rendered as draggable dialog when option is clicked
	 */
	popup?: (...args: any[]) => any;
	/**
	 * Is popup draggable
	 */
	isDraggable?: boolean;
	/**
	 * Always fetch values when menu opens
	 */
	alwaysRefresh?: boolean;
	/**
	 * Function that returns one or many records that will be appened to the result list
	 */
	customOptions?: (...args: any[]) => any;
}

export default class Lookup extends React.Component<LookupProps, any> {
	render(): JSX.Element;
}
