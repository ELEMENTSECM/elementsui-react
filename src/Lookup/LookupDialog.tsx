import * as React from "react";
import { createPortal } from "react-dom";
import Draggable from "react-draggable";
import _ from "lodash";
import classNames from "classnames";
import { isDescendant } from "../utils";

export interface LookupDialogProps {
	/**
	 * Callback executed when a dialog is closed
	 */
	close: (event: any) => any;
	/**
	 * Abolute or fixed positioning for a dialog
	 */
	position: { x: number; y: number };
	/**
	 * Enabled drag n drop for a diealog
	 */
	isDraggable?: boolean;
	/**
	 * Labelled by Id
	 */
	ariaLabelledBy?: string;
	/**
	 * Draggable handle selector
	*/
	dragHandle?: string;
	/**
	 * Portal target id
	 */
	portalTarget?: string;
	/**
	 * Popup container class name
	 */
	className?: string;
	/** 
	 * clickoutside props 
	 */
	[key: string]: any;
}

export default class LookupDialog extends React.Component<LookupDialogProps> {
	static defaultProps = {
		dragHandle: ".popup-container",
		className: "popup-container dialog"
	};
	draggableContainer = React.createRef() as React.MutableRefObject<HTMLElement>;
	containerRef = React.createRef<HTMLDivElement>();

	componentDidMount() {
		if (this.props.isDraggable) {
			this.draggableContainer.current = (this.props.portalTarget
				? this.props.portalTarget.charAt(0) === "."
					? document.getElementsByClassName(this.props.portalTarget.substr(1))[0]
					: document.getElementById(this.props.portalTarget.substr(1))
				: document.body) as HTMLElement;
		}
		this.focusFirstInput();
	}

	handleClickOutside = event => {
		this.props.close(event);
	};

	focusFirstInput = () => {
		const input = _.first(
			this.containerRef.current!.querySelectorAll("input"),
			input => input.type !== "hidden"
		);
		input && input.focus();
	};

	onKeyDown = e => {
		const eventKey = e.key;
		switch (eventKey) {
			case "Escape":
				this.props.close(e);
				break;
			case "Tab":
				if (!isDescendant(this.containerRef.current, e.target)) {
					e.preventDefault();
					this.focusFirstInput();
				}
				break;
		}
	};

	render() {
		const {
			close,
			eventTypes,
			outsideClickIgnoreClass,
			preventDefault,
			stopPropagation,
			disableOnClickOutside,
			enableOnClickOutside,
			position,
			isDraggable,
			dragHandle,
			children,
			portalTarget,
			ariaLabelledBy,
			...rest
		} = this.props;
		const defaultPosition = {
			x: position.x,
			y: position.y + 5
		};

		if (isDraggable) {
			const DraggableDialog = (
				<Draggable handle={dragHandle} defaultPosition={defaultPosition}>
					<div
						ref={this.containerRef}
						role="dialog"
						onKeyDown={this.onKeyDown}
						aria-labelledby={ariaLabelledBy}
						style={{
							minWidth: 100,
							minHeight: 100,
							background: "#fff",
							zIndex: 99999,
							boxShadow: "2px 7px 24px 2px rgba(144,143,143,.8)",
							opacity: 1
						}}
						{...rest}
					>
						{children}
					</div>
				</Draggable>
			);
			return this.draggableContainer.current
				? createPortal(DraggableDialog, this.draggableContainer.current)
				: DraggableDialog;
		}

		return (
			<React.Fragment>
				<div className="popup-overlay" style={{ zIndex: 1055 }} />
				<div
					className={classNames(
						"popup-container ignore-react-onclickoutside",
						this.props.className
					)}
					ref={this.containerRef}
					onKeyDown={this.onKeyDown}
					role="dialog"
					aria-labelledby={ariaLabelledBy}
					style={{
						top: 0,
						left: 0,
						opacity: 1,
						zIndex: 1056
					}}
				>
					{children}
				</div>
			</React.Fragment>
		);
	}
}
