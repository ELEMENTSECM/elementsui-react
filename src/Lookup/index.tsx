export { default, LookupProps } from "./Lookup";
export { default as LookupDialog, LookupDialogProps } from "./LookupDialog";
