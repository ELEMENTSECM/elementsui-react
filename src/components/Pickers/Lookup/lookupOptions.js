import _ from "lodash";

function freetextValue(search) {
	return [
		{
			value: search,
			label: search,
			custom: true
		}
	];
}

function addFreetextValue(search) {
	if (_.isNil(search) || _.isEmpty(search)) {
		return [];
	}

	return freetextValue(search);
}

export function allOptions(currentOptionsOptions, customOptions, includeFreetextValues, search) {
	let allOptions = [];

	if (_.isArray(currentOptionsOptions)) {
		allOptions = allOptions.concat(currentOptionsOptions);
	}

	if (_.isArray(customOptions)) {
		allOptions = allOptions.concat(customOptions);
	}

	if (includeFreetextValues) {
		allOptions = allOptions.concat(addFreetextValue(search));
	}

	// Note the priority here:
	// If an option exist in both lookup + custom lookups + freetext - it will only show the lookup version
	// If the option exists in custom lookups + freetext - it will only show the custom lookup version
	const allOptionsWithoutDuplicates = _.uniqBy(allOptions, "value");

	return allOptionsWithoutDuplicates;
}
