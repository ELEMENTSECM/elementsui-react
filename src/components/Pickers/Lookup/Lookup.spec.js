import * as React from "react";
import { shallow } from "enzyme";
import Lookup from "./index";

describe("Lookup", () => {
	const queryProvider = filter => filter;
	const renderOption = lookupValue => lookupValue.Title;
	const values = [
		{
			Id: 1,
			Title: "Title 1"
		},
		{
			Id: 2,
			Title: "Title 2"
		}
	];

	const options = values.map(x => ({
		value: x.Id,
		label: x.Title
	}));

	let wrapper;
	let component;

	describe("with options property", () => {
		beforeEach(() => {
			wrapper = shallow(
				<Lookup id="testLookup" queryProvider={queryProvider} renderOption={renderOption} options={options} />
			);
			component = wrapper.instance();
		});

		test("should be displayed correctly", () => {
			expect(wrapper).toMatchSnapshot();
		});

		test("should update state with options", () => {
			expect(component.state["optionsCache"][""]["options"]).toEqual(options);
		});

		test("should update state on 'onMenuOpen'", async () => {
			const mockLoadOptions = jest.spyOn(component, "loadOptions");
			await wrapper.find("Select").simulate("menuOpen");
			expect(component.state["menuIsOpen"]).toBeTruthy();
			expect(mockLoadOptions).not.toBeCalled();
		});

		test("should update state on 'onMenuClose'", () => {
			component.state["menuIsOpen"] = true;
			wrapper.find("Select").simulate("menuClose");
			expect(component.state["menuIsOpen"]).toBeFalsy();
		});

		test("should update state with options", () => {
			expect(component.state["optionsCache"][""]["options"]).toEqual(options);
		});

		test("should update state on 'onInputChange' and call 'loadOptions'", async () => {
			const filter = "test";
			const mockLoadOptions = jest.spyOn(component, "loadOptions");
			await wrapper.find("Select").simulate("inputChange", filter);
			expect(component.state["search"]).toEqual(filter);
			expect(mockLoadOptions).toBeCalled();
		});

		test("should avoid calling 'loadOptions' on 'onInputChange'", async () => {
			const mockLoadOptions = jest.spyOn(component, "loadOptions");
			await wrapper.find("Select").simulate("inputChange", "");
			expect(mockLoadOptions).not.toBeCalled();
		});

		test("should call 'loadOptions' on 'onMenuScrollToBottom'", async () => {
			const mockLoadOptions = jest.spyOn(component, "loadOptions");
			await wrapper.find("Select").simulate("menuScrollToBottom");
			expect(mockLoadOptions).toBeCalled();
		});

		test("should skip loading if there are no new items", async () => {
			const mockLoad = jest.spyOn(component, "load");
			component.state["optionsCache"][""]["hasMore"] = false;
			await component.loadOptions();
			expect(mockLoad).not.toBeCalled();
		});
	});

	describe("without options property", () => {
		beforeEach(() => {
			wrapper = shallow(<Lookup id="testLookup" queryProvider={queryProvider} renderOption={renderOption} />);
			component = wrapper.instance();
		});

		test("should initialize optionsCache with initial cache", () => {
			expect(component.state["optionsCache"]).toEqual({
				"": { hasMore: true, isLoading: false, options: [], values: [], isTriedLoad: false }
			});
		});

		test("should update state on 'onMenuOpen' and call 'loadOptions'", async () => {
			const mockLoadOptions = jest.spyOn(component, "loadOptions");
			await wrapper.find("Select").simulate("menuOpen");
			expect(component.state["menuIsOpen"]).toBeTruthy();
			expect(mockLoadOptions).toBeCalled();
		});

		test("should load options", async () => {
			component.load = jest.fn(() => ({
				options,
				values
			}));
			await component.loadOptions();
			expect(component.state["optionsCache"][""]["options"]).toEqual(options);
		});
	});

	describe("with fullObjectValue property", () => {
		const mockOnChange = jest.fn();
		beforeEach(() => {
			wrapper = shallow(
				<Lookup
					id="testLookup"
					queryProvider={queryProvider}
					renderOption={renderOption}
					onChange={mockOnChange}
					fullObjectValue={true}
				/>
			);
			component = wrapper.instance();
		});

		test("should return full object value", async () => {
			component.load = jest.fn(() => ({
				options,
				values
			}));
			await component.loadOptions();
			expect(component.state["optionsCache"][""]["options"]).toEqual(options);
			expect(component.state["optionsCache"][""]["values"]).toEqual(values);
		});
	});
});
