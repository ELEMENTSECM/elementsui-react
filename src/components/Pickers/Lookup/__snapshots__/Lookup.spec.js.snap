// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lookup with options property should be displayed correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Lookup
    delay={300}
    id="testLookup"
    idSelector={[Function]}
    isClearable={true}
    isMulti={false}
    menuPlacement="bottom"
    options={
      Array [
        Object {
          "label": "Title 1",
          "value": 1,
        },
        Object {
          "label": "Title 2",
          "value": 2,
        },
      ]
    }
    pageSize={10}
    queryProvider={[Function]}
    renderOption={[Function]}
    selectRef={[Function]}
    styles={
      Object {
        "control": [Function],
        "fullObjectValue": false,
        "menu": [Function],
        "menuPortal": [Function],
        "placeholder": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Select
          backspaceRemovesValue={true}
          blurInputOnSelect={true}
          captureMenuScroll={false}
          className={undefined}
          closeMenuOnScroll={false}
          closeMenuOnSelect={true}
          components={Object {}}
          controlShouldRenderValue={true}
          escapeClearsValue={false}
          filterOption={[Function]}
          formatGroupLabel={[Function]}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          id="testLookup"
          inputId={undefined}
          inputValue=""
          isClearable={true}
          isDisabled={false}
          isLoading={false}
          isMulti={false}
          isOptionDisabled={[Function]}
          isRtl={false}
          isSearchable={true}
          loadOptions={[Function]}
          loadingMessage={[Function]}
          maxMenuHeight={300}
          menuIsOpen={false}
          menuPlacement="bottom"
          menuPortalTarget={<body />}
          menuPosition="absolute"
          menuShouldBlockScroll={false}
          menuShouldScrollIntoView={true}
          minMenuHeight={140}
          noOptionsMessage={[Function]}
          onChange={[Function]}
          onInputChange={[Function]}
          onMenuClose={[Function]}
          onMenuOpen={[Function]}
          onMenuScrollToBottom={[Function]}
          openMenuOnClick={true}
          openMenuOnFocus={true}
          options={
            Array [
              Object {
                "label": "Title 1",
                "value": 1,
              },
              Object {
                "label": "Title 2",
                "value": 2,
              },
            ]
          }
          pageSize={5}
          placeholder="Select..."
          screenReaderStatus={[Function]}
          styles={
            Object {
              "control": [Function],
              "fullObjectValue": false,
              "menu": [Function],
              "menuPortal": [Function],
              "placeholder": [Function],
            }
          }
          tabIndex="0"
          tabSelectsValue={true}
          theme={undefined}
          value={undefined}
        />,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backspaceRemovesValue": true,
          "blurInputOnSelect": true,
          "captureMenuScroll": false,
          "className": undefined,
          "closeMenuOnScroll": false,
          "closeMenuOnSelect": true,
          "components": Object {},
          "controlShouldRenderValue": true,
          "escapeClearsValue": false,
          "filterOption": [Function],
          "formatGroupLabel": [Function],
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "id": "testLookup",
          "inputId": undefined,
          "inputValue": "",
          "isClearable": true,
          "isDisabled": false,
          "isLoading": false,
          "isMulti": false,
          "isOptionDisabled": [Function],
          "isRtl": false,
          "isSearchable": true,
          "loadOptions": [Function],
          "loadingMessage": [Function],
          "maxMenuHeight": 300,
          "menuIsOpen": false,
          "menuPlacement": "bottom",
          "menuPortalTarget": <body />,
          "menuPosition": "absolute",
          "menuShouldBlockScroll": false,
          "menuShouldScrollIntoView": true,
          "minMenuHeight": 140,
          "noOptionsMessage": [Function],
          "onChange": [Function],
          "onInputChange": [Function],
          "onMenuClose": [Function],
          "onMenuOpen": [Function],
          "onMenuScrollToBottom": [Function],
          "openMenuOnClick": true,
          "openMenuOnFocus": true,
          "options": Array [
            Object {
              "label": "Title 1",
              "value": 1,
            },
            Object {
              "label": "Title 2",
              "value": 2,
            },
          ],
          "pageSize": 5,
          "placeholder": "Select...",
          "screenReaderStatus": [Function],
          "styles": Object {
            "control": [Function],
            "fullObjectValue": false,
            "menu": [Function],
            "menuPortal": [Function],
            "placeholder": [Function],
          },
          "tabIndex": "0",
          "tabSelectsValue": true,
          "theme": undefined,
          "value": undefined,
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Select
            backspaceRemovesValue={true}
            blurInputOnSelect={true}
            captureMenuScroll={false}
            className={undefined}
            closeMenuOnScroll={false}
            closeMenuOnSelect={true}
            components={Object {}}
            controlShouldRenderValue={true}
            escapeClearsValue={false}
            filterOption={[Function]}
            formatGroupLabel={[Function]}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            id="testLookup"
            inputId={undefined}
            inputValue=""
            isClearable={true}
            isDisabled={false}
            isLoading={false}
            isMulti={false}
            isOptionDisabled={[Function]}
            isRtl={false}
            isSearchable={true}
            loadOptions={[Function]}
            loadingMessage={[Function]}
            maxMenuHeight={300}
            menuIsOpen={false}
            menuPlacement="bottom"
            menuPortalTarget={<body />}
            menuPosition="absolute"
            menuShouldBlockScroll={false}
            menuShouldScrollIntoView={true}
            minMenuHeight={140}
            noOptionsMessage={[Function]}
            onChange={[Function]}
            onInputChange={[Function]}
            onMenuClose={[Function]}
            onMenuOpen={[Function]}
            onMenuScrollToBottom={[Function]}
            openMenuOnClick={true}
            openMenuOnFocus={true}
            options={
              Array [
                Object {
                  "label": "Title 1",
                  "value": 1,
                },
                Object {
                  "label": "Title 2",
                  "value": 2,
                },
              ]
            }
            pageSize={5}
            placeholder="Select..."
            screenReaderStatus={[Function]}
            styles={
              Object {
                "control": [Function],
                "fullObjectValue": false,
                "menu": [Function],
                "menuPortal": [Function],
                "placeholder": [Function],
              }
            }
            tabIndex="0"
            tabSelectsValue={true}
            theme={undefined}
            value={undefined}
          />,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backspaceRemovesValue": true,
            "blurInputOnSelect": true,
            "captureMenuScroll": false,
            "className": undefined,
            "closeMenuOnScroll": false,
            "closeMenuOnSelect": true,
            "components": Object {},
            "controlShouldRenderValue": true,
            "escapeClearsValue": false,
            "filterOption": [Function],
            "formatGroupLabel": [Function],
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "id": "testLookup",
            "inputId": undefined,
            "inputValue": "",
            "isClearable": true,
            "isDisabled": false,
            "isLoading": false,
            "isMulti": false,
            "isOptionDisabled": [Function],
            "isRtl": false,
            "isSearchable": true,
            "loadOptions": [Function],
            "loadingMessage": [Function],
            "maxMenuHeight": 300,
            "menuIsOpen": false,
            "menuPlacement": "bottom",
            "menuPortalTarget": <body />,
            "menuPosition": "absolute",
            "menuShouldBlockScroll": false,
            "menuShouldScrollIntoView": true,
            "minMenuHeight": 140,
            "noOptionsMessage": [Function],
            "onChange": [Function],
            "onInputChange": [Function],
            "onMenuClose": [Function],
            "onMenuOpen": [Function],
            "onMenuScrollToBottom": [Function],
            "openMenuOnClick": true,
            "openMenuOnFocus": true,
            "options": Array [
              Object {
                "label": "Title 1",
                "value": 1,
              },
              Object {
                "label": "Title 2",
                "value": 2,
              },
            ],
            "pageSize": 5,
            "placeholder": "Select...",
            "screenReaderStatus": [Function],
            "styles": Object {
              "control": [Function],
              "fullObjectValue": false,
              "menu": [Function],
              "menuPortal": [Function],
              "placeholder": [Function],
            },
            "tabIndex": "0",
            "tabSelectsValue": true,
            "theme": undefined,
            "value": undefined,
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
